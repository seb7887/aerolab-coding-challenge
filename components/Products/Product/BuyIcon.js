import React from 'react';

const BuyIcon = () => <img src='/static/images/buy-orange.svg' alt='Buy' />;

export default BuyIcon;
