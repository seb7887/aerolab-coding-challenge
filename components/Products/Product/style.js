import styled from 'styled-components';

export const Container = styled.div`
  text-align: center;
  box-shadow: ${props => props.theme.productShadow};
  min-height: 150px;
  min-width: 200px;
  transition: all 0.3s;

  &:hover {
    transform: translateY(-1rem);
  }
`;

export const NeedCoins = styled.div`
  font-size: 0.8rem;
  font-weight: 600;
  letter-spacing: 0.5px;
  text-align: center;
  padding: 8px 11px;
  border-radius: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  transition: all 0.3s;
  background-color: ${props => props.theme.secondaryBg};
`;

export const Img = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

export const OverlayBlock = styled.div`
  display: flex;
  flex-direction: column;
`;

export const RedeemButton = styled.a`
  margin: 1rem 2rem;
  font-size: 1rem;
  font-weight: 600;
  letter-spacing: 0.5px;
  text-align: center;
  padding: 9px 12px;
  border-radius: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  transition: all 0.3s;
  background-color: ${props => props.theme.bg};
  cursor: pointer;
`;

export const Content = styled.div`
  position: relative;
  padding: 1.5rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  letter-spacing: 1px;
  transition: all 0.5s;

  .buy {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
  }

  .description {
    display: flex;
    flex-direction: column;
    border-top: 1px solid ${props => props.theme.grey};
    padding-top: 1rem;

    .name {
      font-weight: 600;
    }
  }

  .overlay {
    display: none;
    margin: -1.5rem;
    padding: 1.35rem 1.2rem;
    z-index: 2;
    position: absolute;
    width: 100%;
    height: 100%;
    background-color: rgba(255, 111, 5, 0.8);
    transition: all 0.5s;
  }

  &:hover {
    .overlay {
      display: flex;
      flex-direction: column;
      justify-content: space-between;

      .cost-block {
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: center;

        .cost {
          font-size: 2rem;
          font-weight: bold;
          color: ${props => props.theme.white};
          margin-right: 0.5rem;
        }
      }
    }
  }
`;
