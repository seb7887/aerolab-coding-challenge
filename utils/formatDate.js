export default function formatDate(value) {
  const getNumber = number => (number < 10 ? `0${number}` : String(number));
  const date = new Date(value);

  return `${getNumber(date.getMonth() + 1)}/${getNumber(
    date.getDate()
  )}/${getNumber(date.getFullYear())} ${getNumber(
    date.getUTCHours()
  )}:${getNumber(date.getUTCMinutes())}`;
}
