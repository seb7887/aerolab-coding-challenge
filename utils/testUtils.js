import faker from 'faker';

export const fakeUser = {
  name: 'John Kite',
  points: '7000'
};

export const mockCategories = [
  'all',
  'gaming',
  'laptops',
  'tablets',
  'cameras',
  'audio',
  'tv',
  'drones',
  'accessories',
  'home'
];

export const fakeProduct = {
  _id: 7,
  name: 'iPhone',
  category: 'Phones',
  cost: 800,
  img: {
    url: '/static/image/image.png'
  }
};

export const generateProduct = () => {
  const newProduct = {
    _id: faker.random.uuid(),
    name: faker.commerce.productName(),
    category: faker.commerce.productAdjective(),
    cost: 800,
    img: {
      url: faker.image.imageUrl()
    }
  };
  return newProduct;
};

export const fakeProducts = [fakeProduct];

export const generateProducts = quantity => {
  let products = [];

  for (let i = 0; i < quantity; i++) {
    const newProduct = generateProduct();
    products.push(newProduct);
  }

  return products;
};
