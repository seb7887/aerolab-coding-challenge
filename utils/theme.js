const black = '#000';
const white = '#fff';
const cyan = '#0ad4fa';
const orange = '#ff6e05';
const red = 'red';
const green = 'green';

const bgLight = '#e9ebf1';
const bgDark = '#212531';
const secondaryBgLigth = 'rgba(0, 0, 0, 0.1)';
const secondaryBgDark = 'rgba(255, 255, 255, 0.1)';
const containerBgLight = '#fff';
const containerBgDark = '#000';
const boxShadowLight = '0px 0px 3px 0px rgba(0, 0, 0, 0.1)';
const boxShadowDark = 'rgba(31, 31, 31, 0.1)';
const productShadowLight = '1px 4px 5px 1px rgba(0, 0, 0, 0.1)';
const productShadowDark = '1px 4px 5px 1px rgba(255, 255, 255, 0.1)';

const primaryLight = '#616161';
const primaryDark = '#f8f8f8';

const secondaryLight = black;
const secondaryDark = white;
const greyLight = '#e0e0e0';
const greyDark = '#a3a3a3';

const theme = {
  light: {
    black,
    white,
    cyan,
    orange,
    red,
    green,
    grey: greyLight,
    bg: bgLight,
    secondaryBg: secondaryBgLigth,
    containerBg: containerBgLight,
    boxShadow: boxShadowLight,
    productShadow: productShadowLight,
    primary: primaryLight,
    secondary: secondaryLight
  },
  dark: {
    black,
    white,
    cyan,
    orange,
    red,
    green,
    grey: greyDark,
    bg: bgDark,
    secondaryBg: secondaryBgDark,
    containerBg: containerBgDark,
    boxShadow: boxShadowDark,
    productShadow: productShadowDark,
    primary: primaryDark,
    secondary: secondaryDark
  }
};

export default theme;
