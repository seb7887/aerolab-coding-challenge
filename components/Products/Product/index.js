import React from 'react';

import BuyIcon from './BuyIcon';
import Coin from '../../shared/Coin';

import {
  Container,
  Img,
  Content,
  OverlayBlock,
  RedeemButton,
  NeedCoins
} from './style';

const Product = ({ product, userPoints }) => (
  <Container>
    <Content>
      <div className='buy'>
        {userPoints < product.cost ? (
          <NeedCoins>
            You need {product.cost - userPoints} <Coin />
          </NeedCoins>
        ) : (
          <BuyIcon color='orange' />
        )}
      </div>
      <Img src={product.img.url} alt={product.name} />
      <div className='description'>
        <span className='category'>{product.category}</span>
        <span className='name'>{product.name}</span>
      </div>
      {userPoints >= product.cost && (
        <div className='overlay'>
          <div className='buy'>
            <BuyIcon color='white' />
          </div>
          <OverlayBlock>
            <div className='cost-block'>
              <span className='cost'>{product.cost}</span>
              <Coin />
            </div>
            <RedeemButton>Redeem Now</RedeemButton>
          </OverlayBlock>
        </div>
      )}
    </Content>
  </Container>
);

export default Product;
