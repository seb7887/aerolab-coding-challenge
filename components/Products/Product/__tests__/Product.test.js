import { render, fireEvent } from 'react-testing-library';

import Product from '../index';

import { generateProduct, fakeProduct } from '../../../../utils/testUtils';

const mockRedeemProduct = jest.fn();

let userPoints = 300;
const product = generateProduct();

describe('Product', () => {
  it('should display product description', () => {
    const { getByTestId } = render(
      <Product
        product={product}
        userPoints={userPoints}
        redeemProduct={mockRedeemProduct}
      />
    );

    expect(getByTestId('category').textContent).toBe(product.category);
    expect(getByTestId('name').textContent).toBe(product.name);
  });

  it('should display coins needed if user cannot afford it', () => {
    const pointsToReach = product.cost - userPoints;
    const { getByTestId } = render(
      <Product
        product={product}
        userPoints={userPoints}
        redeemProduct={mockRedeemProduct}
      />
    );

    expect(getByTestId('need-coins')).toBeTruthy();
    expect(getByTestId('need-coins').textContent).toContain(
      `You need ${pointsToReach}`
    );
  });

  it('should display product cost and redeem button if user can affort it', () => {
    userPoints = 900;
    const { getByTestId } = render(
      <Product
        product={product}
        userPoints={userPoints}
        redeemProduct={mockRedeemProduct}
      />
    );

    expect(getByTestId('cost').textContent).toBe(product.cost.toString());
    expect(getByTestId('redeem-button').textContent).toBe('Redeem Now');
  });

  it('should redeem if button is pressed', () => {
    userPoints = 900;
    const { getByTestId } = render(
      <Product
        product={product}
        userPoints={userPoints}
        redeemProduct={mockRedeemProduct}
      />
    );

    fireEvent.click(getByTestId('redeem-button'));

    expect(mockRedeemProduct).toHaveBeenCalledTimes(1);
  });

  it('renders and matches snapshot', () => {
    userPoints = 900;
    const { asFragment } = render(
      <Product
        product={fakeProduct}
        userPoints={userPoints}
        redeemProduct={mockRedeemProduct}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
