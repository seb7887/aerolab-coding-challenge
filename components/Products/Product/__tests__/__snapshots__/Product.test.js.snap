// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="style__Container-g760h8-0 cRZAeB"
    data-testid="product"
  >
    <div
      class="style__Content-g760h8-5 cyJBbK"
    >
      <div
        class="buy"
      >
        <img
          alt="Buy"
          src="/static/images/buy-orange.svg"
        />
      </div>
      <img
        alt="iPhone"
        class="style__Img-g760h8-2 frYZGq"
        src="/static/image/image.png"
      />
      <div
        class="description"
      >
        <span
          class="category"
          data-testid="category"
        >
          Phones
        </span>
        <span
          class="name"
          data-testid="name"
        >
          iPhone
        </span>
      </div>
      <div
        class="overlay"
      >
        <div
          class="buy"
        >
          <img
            alt="Buy"
            src="/static/images/buy-white.svg"
          />
        </div>
        <div
          class="style__OverlayBlock-g760h8-3 lladji"
        >
          <div
            class="cost-block"
          >
            <span
              class="cost"
              data-testid="cost"
            >
              800
            </span>
            <img
              alt="Coin"
              class="Coin__Img-sc-1i19xvt-0 bcGCMK"
              src="/static/images/coin.svg"
            />
          </div>
          <a
            class="style__RedeemButton-g760h8-4 cUoLRy"
            data-testid="redeem-button"
          >
            Redeem Now
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
